server/db-middleware/test-api.js

```diff
@@ -3,35 +3,35 @@ const {
   Request,
   REQUEST_STATUS_APPROVED,
   REQUEST_STATUS_PENDING,
+  REQUEST_STATUS_REJECTED,
 } = require('../models/testing/test-request');
+const { TestAssignment } = require('../models/testing/test-assignment');
+const { TagAttachment } = require('../models/testing/tag-attachment');
 
 const apiModule = {};
 
-apiModule.getStudentRequestsWithStatus = function (userId, statusesToFind) {
+const checkRequestsForSections = function (sectId, studentId) {
   return Request
-    .find({ userId, status: { $in: statusesToFind } });
-};
-
-apiModule.getSectionByRequestId = function (requestId) {
-  return Request
-    .findById(requestId)
-    .populate('sectionId')
-    .then(request => request.sectionId);
+    .countDocuments({
+      userId: studentId,
+      sectionId: sectId,
+      status: { $in: [REQUEST_STATUS_APPROVED, REQUEST_STATUS_PENDING] },
+    })
+    .then((count) => {
+      if (!count) {
+        return Section.findById(sectId);
+      }
+    });
 };
 
-apiModule.acceptableSectionsToRequest = function (userId) {
-  let result = [];
+apiModule.acceptableSectionsToRequest = function (studentId) {
   return Section
     .find()
-    .then((sections) => {
-      result = sections;
-      return apiModule.getStudentRequestsWithStatus(userId, [REQUEST_STATUS_APPROVED, REQUEST_STATUS_PENDING]);
-    })
-    .then(allUnacceptableRequests => Promise.all(allUnacceptableRequests.map(el => this.getSectionByRequestId(el))))
-    .then(checking => result.filter(object => checking.every((element) => {
-      if (element.id === object.id) { return false; }
+    .then(sections => Promise.all(sections.map(el => checkRequestsForSections(el.id, studentId))))
+    .then(sections => sections.filter((object) => {
+      if (object == null) return false;
```

Comment:
Be careful using ==, it has pretty unobvious behavior.
Also do you really need to compare with null? I suppose if (object) will aslo work.

Reply:
Boolean (object) is always true.
Also it should be there to specify, that variable was not defined.